From 3ed85db888114d1b2d4a407edf64660c7250098a Mon Sep 17 00:00:00 2001
From: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
Date: Mon, 20 May 2024 16:12:09 +0300
Subject: [PATCH] Client: Try harder to guess the transient parent window

xdg_popup requires a parent surface. On the other hand, Qt is quite
permissive, it allows showing a popup window without specifying the
transient parent window. Things have been slowly changing and more
controls specify the transient parent, but there are still apps that
don't do it. For that particular reason, QWaylandWindow has some
heuristics to guess the transient parent.

Currently, that heuristic works as follows:

- If QWindow::transientParent() is set, use that
- otherwise use the focus window

However, this heuristic doesn't consider that a window can be inactive.
For example, user hovers an element in the window while the window has
no focus.

This change proposes a heuristic that considers the last interacted
input device, i.e.

- if QWindow::transientParent() is specified, prefer that
- if there's a "top" popup, use it
- otherwise check the last interacted window

With the proposed heuristic, tooltips are less likely to be backed by
xdg-toplevel in inactive apps.

Change-Id: I7e6a607dc38f9e8adce316d3540a9bd9c37cfed4
Reviewed-by: David Edmundson <davidedmundson@kde.org>
(cherry picked from commit 911b66dbd66dbad96e5f07a8b9b7ca56cf3b81b9)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/client/qwaylandwindow.cpp | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index 42fb1312d..251a4ef3b 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -1153,8 +1153,13 @@ QWaylandWindow *QWaylandWindow::guessTransientParent() const
     if (auto transientParent = closestShellSurfaceWindow(window()->transientParent()))
         return transientParent;
 
-    if (QGuiApplication::focusWindow() && (window()->type() == Qt::ToolTip || window()->type() == Qt::Popup))
-        return closestShellSurfaceWindow(QGuiApplication::focusWindow());
+    if (window()->type() == Qt::Popup) {
+        if (mTopPopup)
+            return mTopPopup;
+    }
+
+    if (window()->type() == Qt::ToolTip || window()->type() == Qt::Popup)
+        return display()->lastInputWindow();
 
     return nullptr;
 }
-- 
GitLab

